### PR TITLE
Remove the `alloc_jemalloc` crate

### LINKED ISSUE #36963: Switch the default global allocator to System, remove alloc_jemalloc, use jemallocator in rustc
## Updated description

A long time coming, this issue is that we should implement these changes simultaneously:

* Remove the `alloc_jemalloc` crate
* Default allocations for all crate types to `std::alloc::System`. While currently the default for cdylib/staticlib, it's not the default for staticlib/executable
* Add the `jemallocator` crate to rustc, but only rustc
* Long-term, deprecate and remove the `alloc_system` crate

We for the longest time have defaulted to jemalloc as the default allocator for Rust programs. This has been in place since pre-1.0 and the vision was that we'd give programs a by-default faster allocator than what's on the system. Over time, this has not fared well:

* Jemalloc has been disabled on a wide variety of architectures for various reasons, the system allocator seems more reliable.
* Jemalloc, for whatever reason as we ship it, is incompatible with valgrind
* Jemalloc bloats the size of executables by deafult
* Not all Rust programs are bottlenecked on allocations, and those which are can use `#[global_allocator]` to opt-in to a jemalloc-based global allocator (through the `jemallocator` or any other allocator crate).

The compiler, however [still receives a good deal of benefit from using jemalloc](https://perf.rust-lang.org/compare.html?start=74ff7dcb1388e60a613cd6050bcd372a3cc4998b&end=7e7928dc0340d79b404e93f0c79eb4b946c1d669&stat=wall-time) (measured in https://github.com/rust-lang/rust/pull/55202#issuecomment-431514148). If that link is broken, it's basically a blanket across-the-board 8-10% regression in compile time for many benchmarks. (apparently the [max rss also regressed on many benchmarks!](https://perf.rust-lang.org/compare.html?start=74ff7dcb1388e60a613cd6050bcd372a3cc4998b&end=7e7928dc0340d79b404e93f0c79eb4b946c1d669&stat=max-rss)). For this reason, we don't want to remove jemalloc from rustc itself.

The rest of this issue is now going to be technical details about how we can probably get rid of `alloc_jemalloc` while preserving jemalloc in rustc itself. The tier 1 platforms that use `alloc_jemalloc` which this issue will be focused on are:

* x86_64-unknown-linux-gnu
* i686-unknown-linux-gnu
* x86_64-apple-darwin
* i686-apple-darwin

Jemalloc is notably disabled on all Windows platforms (I believe due to our inability to ever get it building over there). Furthermore Jemalloc is enabled on some linux platforms but I think ended up basically being disabled on all but the above. This I believe narrows the targets we need to design for, as we basically need to keep the above working.

Note that we also have two modes of using jemalloc. In one mode we could actually use jemalloc-specific API functions, like `alloc_jemalloc` does today. We could also use the standard API it has and the support to hook into the standard allocator on these two platforms. It's not been measured (AFAIK) at this time the tradeoff between these two strategies. Note that in any case we want to route LLVM's allocations to jemalloc, so we want to be sure to hook into the default allocator *somehow*.

I believe that this default allocator hooking on Linux works by basically relying on its own symbol `malloc` overriding that in `libc`, routing all memory allocation to jemalloc. I'm personally quite fuzzy on the details for OSX, but I think it has something to do with "zone allocators" and not much to do with symbol names. I think this means we can build jemalloc without symbol prefixes on Linux, and with symbol prefixes on OSX, and we should be able to, using that build, override the default allocator in both situations.

I would propose, first, a "hopefully easy" route to solve this:

* Let's link the compiler to the "system allocator". Let's then, on the four platforms above, link to `jemalloc_sys`, pulling in all of jemalloc itself. This should, with the right build configuration, mean that we're not using jemalloc everywhere in the compiler (just as we're rerouting LLVM we're rerouting the compiler).

I'm testing out the performance of this in https://github.com/rust-lang/rust/pull/55217 and will report back with results. Results are that this is [universally positive almost](https://perf.rust-lang.org/compare.html?start=22cc2ae8057d14e980b7c784e1eb2eee26b59e7d&end=10c95ccfa7a7adc12f4e608621ca29f9b98eed29)! @alexcrichton will make a PR.

Failing this @alexcrichton has ideas for a more invasive solution to use jemalloc-specific API calls in rustc itself, but hopefull that won't be necessary...

## Original Description

@alexcrichton and I have increasingly come to think that Rust should not maintain jemalloc bindings in tree and link it by default. The primary reasons being:
- Being opinionated about the default allocator is against Rust's general philosophy of getting as close to the underlying system as possible. We've removed almost all runtime baggage from Rust except jemalloc.
- Due to breakage we've had to disable jemalloc support on some windows configurations, changing our default allocation characteristics there, and offering different implicit "service levels" on different tier 1 platforms.
- Keeping jemalloc working imposes increased maintenance burden. We support a lot of platforms and jemalloc upgrades sometimes do not work across all of them.
- The build system is complicated by supporting jemalloc on some platforms but not all.

For the sake of consistency and maintenance we'd prefer to just always use the system allocator, and make jemalloc an easy option to enable via the [global allocator](https://github.com/rust-lang/rust/issues/27389) and a jemalloc crate on crates.io.

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit removes the `alloc_jemalloc` crate from the standard library and all related configuration. We will no longer be shipping this unstable crate. Rationale for this is provided on https://github.com/rust-lang/rust/issues/36963 and the many linked issues, but I can inline rationale here if desired!

We currently rely on jemalloc for increased perf in the Rust compiler, however. [This perf run shows](https://perf.rust-lang.org/compare.html?start=74ff7dcb1388e60a613cd6050bcd372a3cc4998b&end=7e7928dc0340d79b404e93f0c79eb4b946c1d669&stat=wall-time) that if we switch to glibc 2.23's allocator that it's slower than jemalloc across many benchmarks. [This perf run, however](https://perf.rust-lang.org/compare.html?start=22cc2ae8057d14e980b7c784e1eb2eee26b59e7d&end=10c95ccfa7a7adc12f4e608621ca29f9b98eed29), shows that if we use `jemalloc-sys` from crates.io then rustc actually gets faster across all benchmarks! (presumably because it has a more recent version of jemalloc than our submodule).

As a result, it's expected that this doesn't regress any code (as it's just removing an unstable crate) and it should actually improve rustc performance because it updates jemalloc.

Closes #36963

### CODE REVIEW

**Comment by lnicola** [comment]
Ah, that might explain https://perf.rust-lang.org/compare.html?start=22cc2ae8057d14e980b7c784e1eb2eee26b59e7d&end=10c95ccfa7a7adc12f4e608621ca29f9b98eed29&stat=max-rss.

**Comment by SimonSapin** [comment]
> We may want to fold the switch to the system allocator as the default into this change as well?

As far as I understand, that switch is already part of this PR.

**Comment by alexcrichton** [comment]
@sfackler ah yeah as @SimonSapin mentioned this PR is doing the switch as well as the update. 

Despite it [looking like we enable the background thread by default](https://github.com/alexcrichton/jemallocator/blob/master/jemalloc-sys/Cargo.toml#L29) in `jemalloc-sys` the build script only actually [disables the background thread if it's turned off](https://github.com/alexcrichton/jemallocator/blob/master/jemalloc-sys/build.rs#L172-L174). @sfackler do you know the right option necessary to enable the background thread so we can test in rustc?

**Comment by sfackler** [comment]
Those options compile in support for the background thread but it's still disabled at runtime by default. I don't think it would have much of an effect on rustc memory usage - my understanding is that it's used to clean up caches on idle threads and that active threads clean themselves up periodically when touching the allocator.

There are a bunch of settings you can poke at to control how aggressively jemalloc caches if you want: http://jemalloc.net/jemalloc.3.html

File: src/librustc_driver/lib.rs [hunk 1]
```diff
@@ -64,6 +64,14 @@ extern crate syntax;
 extern crate syntax_ext;
 extern crate syntax_pos;
 
+// Note that the linkage here should be all that we need, on Linux we're not
+// prefixing the symbols here so this should naturally override our default
+// allocator. On OSX it should override via the zone allocator. We shouldn't
+// enable this by default on other platforms, so other platforms aren't handled
+// here yet.
+#[cfg(feature = "jemalloc-sys")]
+extern crate jemalloc_sys;
```
Comment (gnzlbg) [inline src/librustc_driver/lib.rs]: I get that on these platforms `global_allocator` is not needed, but I still do wonder why this does not use it.

Comment (alexcrichton) [reply src/librustc_driver/lib.rs]: The `#[global_allocator]` attribute is actually incompatible with this due to the way linkage works and such. I can go into details if necessary, but the answer to your question at least is "it's required by the way we build rustc"

**Comment by alexcrichton** [comment]
@bors: retry

After many attempts locally I can't reproduce, so I'm curious if we can get another failure on bors...

File: src/librustc_driver/lib.rs (hunk 1 above)
Comment (gnzlbg) [reply src/librustc_driver/lib.rs]: Ah damn, thanks for the explanation.

**Comment by SimonSapin** [comment]
> There is one pretty significant difference between jemalloc 4 and 5. […]

I *think* that jemalloc-sys 0.1.2 packages exactly the same version of jemalloc that rustc currently uses. (That is, 4.5.0 plus a couple patches https://github.com/jemalloc/jemalloc/compare/4.5.0...rust-lang:rust-2017-05-09)

Maybe this PR should use that, and later we can consider upgrading to jemalloc 5 separately?

```sh
% cd jemallocator/jemalloc-sys
% paste <(git log -p jemalloc|grep Date|grep -v 'Fri Apr 27') <(for c in $(git log -p jemalloc|grep '^+S'|cut -d' ' -f3); (cd jemalloc; echo -n "$c "; git show $c:ChangeLog|head|grep \*))
Date:   Wed May 16 19:16:50 2018 +0200	61efbda7098de6fe64c362d309824864308c36d4 * 5.1.0 (May 4th, 2018)
Date:   Thu May 3 16:37:24 2018 +0200	b8f4c730eff28edee4b583ff5b6ee1fac0f26c27 * 5.0.1 (July 1, 2017)
Date:   Thu May 3 00:23:40 2018 +0200	b8f4c730eff28edee4b583ff5b6ee1fac0f26c27 * 5.0.1 (July 1, 2017)
Date:   Thu Nov 23 19:08:53 2017 +0100	3c8e8db1b57e8101c33b2738213aad07cecbb3d3 * 5.0.1 (July 1, 2017)
Date:   Thu Jul 27 11:07:15 2017 -0300	b6bdb328cf73b67129f7e0a92a513d158763aadb * 4.5.0 (February 28, 2017)
Date:   Mon May 29 21:03:47 2017 +0800	3288e0659c08fb5006f6d6dd4b5675ed0c2c432a * 4.5.0 (February 28, 2017)
Date:   Thu Mar 30 22:15:07 2017 +0800	11bfb0dcf85f7aa92abd30524bb1e42e18d108c6 * 4.1.0 (February 28, 2016)
Date:   Mon Mar 7 13:15:29 2016 -0800	aab1c0a0e0b39825b16673128729ef46310a5da8 * 4.1.0 (February 28, 2016)
Date:   Wed Jan 20 17:03:24 2016 -0800	f84e30927284b0c500ed3eaf09e8e159da20ddaf * 4.0.4 (October 24, 2015)

**Comment by alexcrichton** [comment]
@SimonSapin ah good point! Unfortunately though at the time jemalloc was always built with `--with-jemalloc-prefix`. If this retry doesn't work out though I'll work on getting something published which supports the exact version of jemalloc rustc currently uses.

**Comment by gnzlbg** [comment]
> Despite it looking like we enable the background thread by default in jemalloc-sys the build script only actually disables the background thread if it's turned off. 

This is the intended behavior. What that feature does is turns on/off "background thread support", not whether the background thread is enabled or not. The docs and the feature name could be better - I'll send a PR for that.

Enabling the background thread can be done at compile-time or at run-time. To enable it at compile-time one can: 

* set the `JEMALLOC_SYS_WITH_MALLOC_CONF` environment variable to `JEMALLOC_SYS_WITH_MALLOC_CONF=background_thread:true` when building `jemalloc-sys` (we can probably add a crate feature that does this)
* export `malloc_conf`:

    ```rust
    #[cfg_attr(prefixed, export_name = "_rjem_malloc_conf")]
    #[cfg_attr(not(prefixed), export_name = "malloc_conf")]
    pub static mut malloc_conf: libc::c_char 
        = &b"background_thread:true\0"[0]as *const u8 as *const _;
    ```

AFAICT only users of the final binary can enable it at run-time, by setting the configuration string either in  `/etc/malloc.conf` or in the `MALLOC_CONF` environment variable.

As mentioned, I think it makes sense to rename the current `bg_thread` option to `bg_thread_support`, and to add a new `bg_thread` option that enables the background thread at compile-time, since that is what probably @alexcrichton intended here.

**Comment by kennytm** [comment]
@bors retry

Bors bug.

**Comment by alexcrichton** [comment]
@bors: r=estebank

I've attempted to fix that error by reducing the maximum atomic size on the arm android platform, it looks like they're all using intrinsics anyway...

**Comment by alexcrichton** [comment]
@bors: r-

Ok that's not spurious any more

**Comment by alexcrichton** [comment]
Ok we've got two failing builds so far:

* [A](https://travis-ci.org/rust-lang/rust/jobs/444500939) ([raw](https://api.travis-ci.org/v3/job/444500939/log.txt)) - dist-x86_64-musl
* [B](https://travis-ci.org/rust-lang/rust/jobs/448573473) ([raw](https://api.travis-ci.org/v3/job/448573473/log.txt)) - x86_64-gnu-llvm-5.0

Upon reflection, **neither of these images has jemalloc enabled** with this PR. So... that's quite worrisome. This bug is happening when jemalloc is turned off, not switched to a different version.

Both failures so far have been on the same test [`ui/huge-enum.rs`](https://github.com/rust-lang/rust/blob/05812fa8c588473f03e3fa7cf59cd84f4f37c715/src/test/ui/huge-enum.rs). Both failures seems to produce the right error, but the compiler is exiting in both cases with signal 11 (SIGSEGV) instead of a clean exit with code 1.

The first stack trace:

<details>

```
warning: Could not load shared library symbols for 7 libraries, e.g. /lib/x86_64-linux-gnu/libc.so.6.

Use the "info sharedlibrary" command to see the complete listing.

Do you need "set solib-search-path" or "set sysroot"?

Core was generated by `/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc /checkout/src/tes'.

Program terminated with signal SIGSEGV, Segmentation fault.

#0  0x00007f53f9dbfeaa in llvm::TargetPassConfig::addMachinePasses() () from ./checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/codegen-backends/librustc_codegen_llvm-llvm.so

#0  0x00007f53f9dbfeaa in llvm::TargetPassConfig::addMachinePasses() () from ./checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/codegen-backends/librustc_codegen_llvm-llvm.so

#1  0x00007f53f9bbda0f in llvm::LLVMTargetMachine::addPassesToEmitFile(llvm::legacy::PassManagerBase&, llvm::raw_pwrite_stream&, llvm::raw_pwrite_stream*, llvm::TargetMachine::CodeGenFileType, bool, llvm::MachineModuleInfo*) () from ./checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/codegen-backends/librustc_codegen_llvm-llvm.so

#2  0x00007f53f8ac4354 in LLVMRustWriteOutputFile () from ./checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/codegen-backends/librustc_codegen_llvm-llvm.so

#3  0x00007f53f8978bbc in rustc_codegen_llvm::back::write::write_output_file::h43d626d1bbd593ab () from ./checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/codegen-backends/librustc_codegen_llvm-llvm.so

#4  0x00007f53f896a63d in rustc_codegen_llvm::back::write::codegen::_$u7b$$u7b$closure$u7d$$u7d$::h334d90084152b97a () from ./checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/codegen-backends/librustc_codegen_llvm-llvm.so

#5  0x00007f53f8961c39 in rustc::util::common::time_ext::h1093c0a1462f2cda () from ./checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/codegen-backends/librustc_codegen_llvm-llvm.so

#6  0x00007f53f897b389 in rustc_codegen_llvm::back::write::codegen::h7e2546fe2c85c981 () from ./checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/codegen-backends/librustc_codegen_llvm-llvm.so

#7  0x00007f53f897fbca in rustc_codegen_llvm::back::write::execute_work_item::h34a3780663354f1b () from ./checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/codegen-backends/librustc_codegen_llvm-llvm.so

#8  0x00007f53f893aba2 in std::sys_common::backtrace::__rust_begin_short_backtrace::h95caea62758ed294 () from ./checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/codegen-backends/librustc_codegen_llvm-llvm.so

#9  0x00007f53f8972e49 in std::panicking::try::do_call::hfbe7bba3f27cb8ef () from ./checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/codegen-backends/librustc_codegen_llvm-llvm.so

#10 0x00007f54034da3da in __rust_maybe_catch_panic () at libpanic_unwind/lib.rs:102

#11 0x00007f53f89779f8 in _$LT$F$u20$as$u20$alloc..boxed..FnBox$LT$A$GT$$GT$::call_box::hf492e470c0c53396 () from ./checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/codegen-backends/librustc_codegen_llvm-llvm.so

#12 0x00007f54034cad9e in call_once<(),()> () at liballoc/boxed.rs:682

#13 std::sys_common::thread::start_thread::he4cc3b686d083587 () at libstd/sys_common/thread.rs:24

#14 0x00007f54034a0756 in std::sys::unix::thread::Thread::new::thread_start::hff4201b83401a7d4 () at libstd/sys/unix/thread.rs:90

#15 0x00007f53fd50d6ba in ?? ()

#16 0x0000000000000000 in ?? ()
```

</details>

shows a segfault randomly in LLVM. The second stack trace:

<details>

```
Core was generated by `/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc /checkout/src/tes'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00007f51f8e398ba in ?? ()
#0  0x00007f51f8e398ba in ?? ()
#1  0x0000000000000000 in ?? ()
```

</details>

isn't too helpful...

I've run the compiled compiler for that test tens of thousands of times locally, but I can only get it to exit with a normal status code instead of a segfault.

For lack of a better understanding, the best explanation that I can think of is that the compiler is currently exhibiting UB *today* and is only just now being uncovered. I think the UB is that when the main thread exits there are still child threads running.

I believe this error originates from a codegen thread which means there's probably other threads still running in the background. The first stack trace indicates that to me because if codegen fails we shouldn't write an object file, but the segfault happened while writing an object file! I believe this means that we're forking off multiple codegen modules, and some small ones are still in the process of being worked on when the main codegen threads panics due to this codegen error (what the UI test is testing).

My best guess is that jemalloc just somehow behaves differently than glibc with respect to threads exiting.

Unless anyone else has a suspicion for what's going on here, I'll look into a fix for this.

**Comment by alexcrichton** [comment]
@bors: r=estebank

cc @michaelwoerister, you're likely now interested in the last commit

**Comment by alexcrichton** [comment]
@bors: retry

Docker container needed rebuilding, things timed out

**Comment by jonhoo** [comment]
Just for clarity, does this change *just* affect `rustc`, or does it also affect all crates? Since it closes #36963, does that mean that all crates will now be compiled with the system memory allocator, or are we just moving from `alloc_jemalloc` to `jemalloc-sys`?

File: src/librustc_codegen_llvm/back/write.rs [hunk 2]
```diff
@@ -2464,6 +2504,11 @@ impl OngoingCodegen {
     }
 }
 
+// impl Drop for OngoingCodegen {
```
Comment (gnzlbg) [inline src/librustc_codegen_llvm/back/write.rs]: @alexcrichton what's this?

**Comment by SimonSapin** [comment]
Yes, `System` is now the default on all platforms for all crate types (such as executable). It’s only rustc that this PR makes link to `jemalloc-sys`. I’ve written up an announcement:

* https://internals.rust-lang.org/t/jemalloc-was-just-removed-from-the-standard-library/8759
* https://www.reddit.com/r/rust/comments/9twam5/jemalloc_was_just_removed_from_the_standard/

**Comment by alexcrichton** [comment]
@jonhoo this affects all crates, no crate will be compiled against jemalloc by default now. The compilers we distribute for OSX/Linux will, however, still enable jemalloc (but only the compiler itself)

File: src/librustc_codegen_llvm/back/write.rs (hunk 2 above)
Comment (alexcrichton) [reply src/librustc_codegen_llvm/back/write.rs]: Ah this is a mistake

**Comment by nnethercote** [comment]
[Here](https://perf.rust-lang.org/compare.html?start=757d6cc91a691c8a749a7b0eee09df494ecdcc3b&end=b6e8f9dbdc4f154d799589ce5e21c020c51f2123&stat=instructions:u) are the final perf results from the landing. Almost everything improved by at least 1%, and some were up to 15%. Pretty great!

**Comment by SimonSapin** [comment]
This is most likely from upgrading jemalloc from 4.5 to 5.1, isn’t it?

**Comment by lnicola** [comment]
Are there plans to do something about the memory usage [regression](https://perf.rust-lang.org/compare.html?start=757d6cc91a691c8a749a7b0eee09df494ecdcc3b&end=b6e8f9dbdc4f154d799589ce5e21c020c51f2123&stat=max-rss)?

**Comment by gnzlbg** [comment]
@lnicola the `jemalloc-sys` master branch has re-added support for enabling background threads unconditionally, which might help here - there is also an open issue about gathering statistics of rustc size-class usage which could help here too (https://github.com/alexcrichton/jemallocator/issues/79).

**Pull request merged**